human_aware_rl/imitation/behavior_cloning_tf2.py

```diff
@@ -0,0 +1,358 @@
+                epochs=training_params['epochs'], validation_split=training_params["validation_split"],
+                class_weight=class_weights,
+                verbose=2 if verbose else 0)
+
+    # Save the model
+    save_bc_model(model_dir, model, bc_params)
+
+    return model
+    
+
+
+def save_bc_model(model_dir, model, bc_params):
+    """
+    Saves the specified model under the directory model_dir. This creates three items
+
+        assets/         stores information essential to reconstructing the context and tf graph
+        variables/      stores the model's trainable weights
+        saved_model.pd  the saved state of the model object
+
+    Additionally, saves a pickled dictionary containing all the parameters used to construct this model
+    at model_dir/metadata.pickle
+    """   
+    print("Saving bc model at ", model_dir)
+    model.save(model_dir, save_format='tf')
+    with open(os.path.join(model_dir, "metadata.pickle"), 'wb') as f:
+        pickle.dump(bc_params, f)
+
+
+def load_bc_model(model_dir):
+    """
+    Returns the model instance (including all compilation data like optimizer state) and a dictionary of parameters
+    used to create the model
+    """
+    print("Loading bc model from ", model_dir)
+    model = keras.models.load_model(model_dir)
+    with open(os.path.join(model_dir, "metadata.pickle"), "rb") as f:
+        bc_params = pickle.load(f)
+    return model, bc_params
+
+def evaluate_bc_model(model, bc_params):
+    """
+    Creates an AgentPair object containing two instances of BC Agents, whose policies are specified by `model`. Runs
+    a rollout using AgentEvaluator class in an environment specified by bc_params
+
+    Arguments
+
+        - model (tf.keras.Model)        A function that maps featurized overcooked states to action logits
+        - bc_params (dict)              Specifies the environemnt in which to evaluate the agent (i.e. layout, reward_shaping_param)
+                                            as well as the configuration for the rollout (rollout_length)
+
+    Returns
+
+        - reward (int)                  Total sparse reward achieved by AgentPair during rollout
+    """
+    evaluation_params = bc_params['evaluation_params']
+    mdp_params = bc_params['mdp_params']
+
+    # Get reference to state encoding function used by bc agents, with compatible signature
+    base_env = _get_base_env(bc_params)
+    mlp = get_mlp(base_env)
+    def featurize_fn(state):
+        return base_env.mdp.featurize_state(state, mlp)
+
+    # Wrap Keras models in rllib policies
+    agent_0_policy = BehaviorCloningPolicy.from_model(model, bc_params, stochastic=True)
+    agent_1_policy = BehaviorCloningPolicy.from_model(model, bc_params, stochastic=True)
+
+    # Compute the results of the rollout(s)
+    results = evaluate(evaluation_params, mdp_params, agent_0_policy, agent_1_policy, featurize_fn, featurize_fn)
```

Comment:
Not sure what you mean by "get_rollouts"? The `evaluate` function was just something I wrote out of convenience since I found myself duplicating the code to wrap rllib policies in overcooked agents and calling `AgentEvaluator.evaluate_agent_pair` a bunch. Please let me know if there's something obvious I'm missing here

Reply:
Oh, I wrote this mainly as a reminder to myself. When we talked last time, we had addressed this – the main conclusion was to just try to operate in Agent-class space as much as possible!